flink-java/src/main/java/org/apache/flink/api/java/operators/UnsortedGrouping.java

```diff
@@ -135,10 +136,37 @@ public UnsortedGrouping(DataSet<T> set, Keys<T> keys) {
 	 * @see DataSet
 	 */
 	public ReduceOperator<T> reduce(ReduceFunction<T> reducer) {
+		return reduce(Utils.getCallLocationName(), reducer, CombineHint.OPTIMIZER_CHOOSES);
+	}
+
+	/**
+	 * Applies a Reduce transformation on a grouped {@link DataSet}.<br>
+	 * For each group, the transformation consecutively calls a {@link org.apache.flink.api.common.functions.RichReduceFunction}
+	 *   until only a single element for each group remains.
+	 * A ReduceFunction combines two elements into one new element of the same type.
+	 *
+	 * @param reducer The ReduceFunction that is applied on each group of the DataSet.
+	 * @param strategy The strategy that should be used to execute the combine phase of the reduce.
+	 *                 If {@code null} is given, then the optimizer will pick the strategy.
+	 * @return A ReduceOperator that represents the reduced DataSet.
+	 *
+	 * @see org.apache.flink.api.common.functions.RichReduceFunction
+	 * @see ReduceOperator
+	 * @see DataSet
+	 */
+	public ReduceOperator<T> reduce(ReduceFunction<T> reducer, CombineHint strategy) {
```

Comment:
I don't think we should overload all reduce (or reduce-based) methods in `UnsortedGrouping` but rather add a method `setCombineHint(CombineHint)` to `ReduceOperator`. That way the API doesn't grow too much. This is also more modular once we add a hash-based strategy for the final Reduce.